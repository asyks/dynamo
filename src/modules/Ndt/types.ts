export interface ServerInfo {
  server: string
  path: string
  port: number
}

export type Socketable = ServerInfo | WebSocket

export type PossibleWebSocket = WebSocket | undefined

export const isServerInfo = (socketableObj: Socketable): socketableObj is ServerInfo => {
  // Type guard for PossibleWebsocket constructor
  const testObj = socketableObj as ServerInfo
  return (
    testObj.server !== undefined
    && testObj.port !== undefined
    && testObj.path !== undefined
  )
}
