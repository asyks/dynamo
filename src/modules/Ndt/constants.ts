export enum MessageType {
  COMM_FAILURE = 0,
  SRV_QUEUE = 1,
  MSG_LOGIN = 2,
  TEST_PREPARE = 3,
  TEST_START = 4,
  TEST_MSG = 5,
  TEST_FINALIZE = 6,
  MSG_ERROR = 7,
  MSG_RESULTS = 8,
  MSG_LOGOUT = 9,
  MSG_WAITING = 10,
  MSG_EXTENDED_LOGIN = 11,
}

export enum TestIds {
  TEST_MID = 1,
  TEST_C2S = 2,
  TEST_S2C = 4,
  TEST_SFW = 8,
  TEST_STATUS = 16,
  TEST_META = 32,
}

export const ndtVersion: string = "3.7.0.2"

export const int255HexLiteral: number = 0xFF

export const mLabLocateApiV2Url: URL = new URL(
  "https://locate.measurementlab.net/v2/nearest/ndt/ndt7"
)
