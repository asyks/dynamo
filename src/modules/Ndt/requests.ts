import * as https from 'https'

export const asyncGet = async (url: URL): Promise<Object> => {
  return new Promise((resolve, reject) => {
    let respData: string = ""
    https.get(url.toString(), resp => {
      resp.on("data", (chunk) => { console.log(resp); respData += chunk })
      resp.on("end", () => { console.log(respData); resolve(JSON.parse(respData)) })
    }).on("error", (err) => {
      reject({ error: err.message })
    })
  })
}
