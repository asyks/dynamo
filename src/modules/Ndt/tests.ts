import { serviceUrlPair } from "./types";
import NdtClient from './index'

export const startTest = (serviceUrls: serviceUrlPair): void => {
  console.log(serviceUrls)
  const client = new NdtClient(serviceUrls.upload)

  client.websocket.onopen = (event) => {
    console.log("Opened connection")
    client.login()
  }
}
