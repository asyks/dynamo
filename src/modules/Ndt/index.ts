import { int255HexLiteral, MessageType, ndtVersion } from './constants'
import { PossibleWebSocket, Socketable, isServerInfo, ClientMessage, ServerMessage } from './types'

export interface NdtClientInterface {
  websocket: PossibleWebSocket
}

export default class NdtClient implements NdtClientInterface {
  websocket: PossibleWebSocket

  public constructor(socketableObj: Socketable) {
    if (isServerInfo(socketableObj)) {
      this.websocket = (
        new WebSocket(
          socketableObj.server + socketableObj.path + socketableObj.port
        )
      )
    }
    else {
      this.websocket = socketableObj
    }
    this.websocket.binaryType = "arraybuffer"
  }

  private makeMessageArray(
    messageType: MessageType, messageBody: string
  ): Uint8Array {
    /* Send login message to server.

    Client-side messages have 3 parts:
      - TYPE: a message type integer from 0  - 11
      - LENGTH: total length of the message BODY in 8-bit bytes
      - BODY: variable length content sent to server

     Length (in bytes) = TYPE (1) + LENGTH (2) + BODY = BODY + 3
    */
    const messageArray = new Uint8Array(messageBody.length + 3)

    messageArray[0] = messageType
    // Ensure LENGTH always occupies 2 bytes
    messageArray[1] = (messageBody.length >> 8) & int255HexLiteral
    messageArray[2] = messageBody.length & int255HexLiteral

    for (let i: number = 0; i < messageBody.length; i += 1) {
      messageArray[3 + i] = messageBody.charCodeAt(i)
    }

    return messageArray
  }

  public send(message: ClientMessage): void {
    /* Send message to server. */
    const messageArray = this.makeMessageArray(
      message.type, JSON.stringify(message.body)
    )

    if (this.websocket !== undefined) {
      this.websocket.send(messageArray)
    }
  }

  public login(testIds: number[] = []): void {
    /* Send login message to server.

     - Login message type is MSG_EXTENDED_LOGIN.
     - Send `tests` to the server as a bitwise OR of test ids.
     - By default the TEST_STATUS (16) test is always sent.
    */
    this.send({
      type: MessageType.MSG_EXTENDED_LOGIN,
      body: {
        msg: ndtVersion,
        tests: testIds.reduce((testsNum, testId) => (testsNum | testId), 16)
      }
    })
  }

  public parseMessage(buffer: ArrayBuffer): ServerMessage {
    /* Parse message received from server.

     - Validate the length of the message BODY by comparing to LENGTH
     - If length is valid assign type from TYPE
     - Construct message from BODY
    */
    const messageArray = new Uint8Array(buffer)
    if (
      (messageArray.length - 3) !== ((messageArray[1] << 8) | messageArray[2])
    ) {
      throw new Error('InvalidLengthError')
    }

    return {
      type: messageArray[0],
      body: String.fromCharCode.apply(
        null, Array.from(messageArray.slice(3))
      ),
    }
  }
}
