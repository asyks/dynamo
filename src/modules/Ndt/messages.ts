import { LoginBody, SendBody } from './types'
import { MessageType, int255HexLiteral } from './constants'

/**
 * Ndt Messages have 3 main parts:
 * * TYPE: a message type integer from 0  - 11
 * * LENGTH: total length of the message BODY in 8-bit bytes
 * * BODY: variable length binary message
 * 
 * The data property holds the bytes for the entire message: TYPE, LENGTH, and BODY
 */
export interface BaseMessageInterface {
  type: MessageType
  length: number
  data: Uint8Array
}

export type ClientBody = SendBody | LoginBody

export interface ClientMessageInterface extends BaseMessageInterface {
  body: SendBody | LoginBody
}

export interface ServerMessageInterface extends BaseMessageInterface {
  body: string
}

export class ClientMessage implements ClientMessageInterface {
  type: MessageType
  length: number
  body: ClientBody
  data: Uint8Array

  /**
   * Construct a message for sending to an ndt server.
   * Message bytes:
   * * byte 1 --> TYPE
   * * bytes 2-3 --> LENGTH
   * * bytes 4-END --> BODY
   * @param type The message type id of the message.
   * @param body The body of the message to create.
   */
  public constructor(type: MessageType, body: ClientBody) {
    this.type = type
    this.body = body

    const bodyAsString = JSON.stringify(this.body)
    this.length = bodyAsString.length

    const messageArray = new Uint8Array(bodyAsString.length + 3)
    messageArray[0] = type
    /** Ensure LENGTH always occupies 2 bytes */
    messageArray[1] = (bodyAsString.length >> 8) & int255HexLiteral
    messageArray[2] = bodyAsString.length & int255HexLiteral

    for (let i: number = 0; i < bodyAsString.length; i += 1) {
      messageArray[3 + i] = bodyAsString.charCodeAt(i)
    }
    this.data = messageArray
  }
}

export class ServerMessage implements ServerMessageInterface {
  type: MessageType
  length: number
  body: string
  data: Uint8Array

  /**
   * Construct a message from data recieved from an ndt server.
   * * Validate the length of the message BODY by comparing to LENGTH.
   * * LENGTH in bytes = Message Length - TYPE(length=1) - LENGTH(length=2)
   * @param buffer
   */
  public constructor(buffer: ArrayBuffer) {
    this.data = new Uint8Array(buffer)
    this.length = this.data.length - 3
    if (
      (this.length) !== ((this.data[1] << 8) | this.data[2])
    ) {
      throw new Error('InvalidLengthError')
    }

    this.type = this.data[0]
    this.body = String.fromCharCode.apply(
      null, Array.from(this.data.slice(3))
    )
  }

}
