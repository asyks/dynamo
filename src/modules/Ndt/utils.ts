import { mLabLocateApiV2Url } from './constants'
import { LocateApiV2Response, LocateApiV2Result, serviceUrlPair } from './types'
import * as requests from './requests'

export const parseSecureServiceUrls = (result: LocateApiV2Result): serviceUrlPair => {
  return {
    download: new URL(result.urls["wss:///ndt/v7/download"]),
    upload: new URL(result.urls["wss:///ndt/v7/upload"]),
  }
}

export const getLocateApiData = async (): Promise<LocateApiV2Response> => {
  return await requests.asyncGet(mLabLocateApiV2Url) as LocateApiV2Response
}

export const getServiceUrls = async (resultIndex = 0): Promise<serviceUrlPair> => {
  const serviceData = await getLocateApiData()
  return parseSecureServiceUrls(serviceData.results[resultIndex])
}
