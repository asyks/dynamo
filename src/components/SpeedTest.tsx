import * as React from 'react'

import { getServiceUrls } from '../modules/Ndt/utils'

const SpeedTest: React.FC = () => {
  return (
    <div>
      <div>speed test</div>
      <button onClick={
        async () => { console.log(await getServiceUrls()) }
      }>Connect</button>
    </div>
  )
}

export default SpeedTest
