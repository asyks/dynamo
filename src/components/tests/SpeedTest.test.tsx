import * as React from 'react'
import renderer from 'react-test-renderer'
import { fireEvent, render } from '@testing-library/react'

import * as requests from '../../modules/Ndt/requests'

import SpeedTest from '../SpeedTest'

jest.mock("../../modules/Ndt/requests")

describe("components/SpeedTest", () => {
  let mockedAsyncGet = requests.asyncGet as jest.Mock

  beforeEach(() => {
    mockedAsyncGet.mockResolvedValue({
      results: [{
        urls: {
          "wss:///ndt/v7/download": (
            "wss://ndt-mlab2-lga05.fake.org/ndt/v7/download"
          ),
          "wss:///ndt/v7/upload": (
            "wss://ndt-mlab2-lga05.fake.org/ndt/v7/upload"
          ),
        }
      }]
    })
  })

  test("nominally renders", () => {
    expect(renderer.create(<SpeedTest />)).toMatchSnapshot()
  })

  test("registers button click", () => {
    const { container, getByText } = render(<SpeedTest />)
    fireEvent.click(getByText("Connect"))

    expect(mockedAsyncGet.mock.calls.length).toBe(1)
  })
})
