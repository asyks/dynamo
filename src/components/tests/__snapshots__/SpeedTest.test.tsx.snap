// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/SpeedTest nominally renders 1`] = `
<div>
  <div>
    speed test
  </div>
  <button
    onClick={[Function]}
  >
    Connect
  </button>
</div>
`;
